# BacNex : A comprehensive analysis pipeline for microbiome research

BacNex aims to identify microbiota features differentiating the cohorts, enabling standardized and efficient analysis of microbial datasets and working "end to end" in studies.


## Table of Contents
* [Installation](#intallation)
* [General usage](#general-usage)
  * [Preprocess](#preprocess)
  * [Generate standard input file](#generate-standard-input-file)
  * [Application](#application)
* [Author](#authors)


## Intallation

BacNex is built on the conda environment and supports only Linux x64 systems. Please make sure the [conda](https://docs.conda.io/en/latest/) is installed.

```Shell
$ conda env create -f bacnex_env.yml --name bacnex
$ conda activate bacnex
```

This command will automatically install dependencies of BacNex. The YAML files are in **env**.

#### Automatical installation
1. [HUMAnN 3](https://github.com/biobakery/humann)
2. [MetaPhlAn 4](https://github.com/biobakery/MetaPhlAn)
3. [KneadData](https://github.com/biobakery/kneaddata)
4. [shinyboard](https://github.com/rstudio/shinydashboard)
5. [shinyjs](https://github.com/daattali/shinyjs)
6. [shinycssloaders](https://github.com/daattali/shinycssloaders)
7. [visnetwork](https://github.com/datastorm-open/visNetwork)
8. [shiny](https://shiny.posit.co/)
9. [shinyfeedback](https://github.com/merlinoa/shinyFeedback)
10. [DT](https://github.com/rstudio/DT)
11. [colourpicker](https://github.com/daattali/colourpicker)
12. [ggpubr](https://github.com/kassambara/ggpubr)
13. [uwot](https://github.com/jlmelville/uwot)
14. [patchwork](https://github.com/thomasp85/patchwork)

#### Manual installation
1. [PreLectR](https://github.com/YinchengChen23/PreLectR)

```
$ Rscript -e "install.packages('remotes', repos='https://cloud.r-project.org/')"
$ Rscript -e "remotes::install_github('YinchengChen23/PreLectR')"
```

The software need to be installed by the user.


## Example data
We provides the files for users to quickly demo BacNex as the follows:

1. KO.tsv (HUMAnN3 intergrated data)
2. meta.csv (metadata for demo data)

The example data is deposited in the NCBI BioProject database under project number **PRJNA397112**.


## General usage

To see a list of the available modules, check the help:

```shell
$ bash BacNex.sh --help
```


----
### Build database

```shell
$ bash BacNex.sh make_database \
> -o $OUTDIR
```
The `make_database` module is used to bulid the dependency database of the `preprocess` module. The folder of database `bacnex_db` will be created in target directory.

#### KEGG pathway information map
In BacNex, we have developed a KEGG pathway database map, eliminating the need for users to run this step. The database is built by [KEGGREST](https://github.com/Bioconductor/KEGGREST) (v3.20).


----

### Preprocess

```shell
$ bash BacNex.sh preprocess \
> -i $seq_dir \
> -o $outdir \
> -db $DATABASE \
> --threads 10
```

| Parameters | Description |
|----------|----------|
| -i, --input | The raw fastq directory. |
| -o, --outdir | The output directory. |
| -db, --database| The database directory. |
| -t, --threads | The number of threads. |
| -kd, --kneaddata | Other KneadData options. |
| -hum, --humman | Other HUMAnN options. |
| -mpa, --metaphlan | Other MetaPhlAn options. |

We provide a module to preprocess raw metagenomic sequences. Module `preprocess` used [KneadData (v0.10.0)](https://github.com/biobakery/kneaddata) and [Trimmomatic (v0.39)](http://www.usadellab.org/cms/?page=trimmomatic) to retain microbial sequences and remove contaminants. The functional profile is generated by [HUMAnN 3 (v3.9)](https://github.com/biobakery/humann), which serves as our input file type. You can also preprocess your data independently.

**Issue**:

In certain conditions, KneadData may yeild `ERROR: Unable to find trimmomatic`. Users can manually install Trimmomatic and provide the parameter `--trimmomatic` in `-kd` options.

```
$ bash BacNex.sh preprocess \
> -i $seq_dir \
> -o $outdir \
> -db $DATABASE \
> --threads 10 \
> -kd "/path/trimmomatic"
```

----


### Generate standard input file
```shell
$ python BacNex.py make_table \
> -i $KO_TSV \
> -o $OUTDIR
```

| Parameters | Description |
|----------|----------|
| -i, --input | Path to the HUMAnN format table (ko.tsv). |
| -o, --outdir | The output directory. |

The `make_table` module requires a HUMAnN format table (ko.tsv) as input. It generates the output files which BacNex needs.

#### Three output items will be created:

1. `$OUTDIR/taxa_table.csv`
2. `$OUTDIR/func_table.tsv`
3. `$OUTDIR/taxa_koDict.json`

----


### Application

Call the application
```shell
$ bash BacNex.sh app \
> -w $WORK_DIR \
> -t 10
```

| Parameters | Description |
|----------|----------|
| -w, --work-dir | Path to save all results from the application. |
| -t, --threads | The number of threads. |
| -b, --browser | The selected browser to open application (google-chrome, firefox). (Default: firefox)|


#### *Select the classification task and upload metadata*

PreLect supports four classification tasks, and BacNex can generate the corresponding results. Users can upload metadata in various formats and select a specific task to obtain tailored outcomes.


#### *Diversity analysis*

* Alpha diversity

  Alpha diversity refers to the "within-sample" diversity of a community. BacNex provides visualizations of key metrics, including the Shannon index, Simpson index, and inverse Simpson index.

* Beta diversity

  Beta diversity is the ratio between regional and local species diversity. BacNex provides index including PCoA, NMDS, t-SNE and UMAP.

* Composition

  The taxonomic composition bar plot illustrates the relative abundance of the most prevalent genera in each sample. You can also customize the plot by selecting your preferred taxa level.


#### *Differential taxa*
Module `Differential taxa` employ [PreLect algorithms](https://github.com/YinchengChen23/PreLectR), a feature selection method combining logistic regression with L1 regularization and integrating an inverted prevalence penalty, to discriminate between two cohorts and reveal pathogens and probiotics.

![alt](img/img1.png?raw=true)
**Fig 1.** BCE loss and prevalence at different lambda. Optional lamda (dotted line) is selected at the inflection point of the BCE loss curve.


#### *Network construction*

The networks are constructed based on taxonomy genomic similarity, with Fisher's exact test employed to evaluate the overlap of KOs between pairs of taxa. Taxa pairs are considered connected if the FDR is below a specified threshold, and the odds ratio serves as the weight for the corresponding edge.

![alt](img/img2.png?raw=true)
**Fig 2.** Schematic of taxa-taxa genomic similarity network


#### *Network filter*

After constructing the global genomic similarity network, users can identify important sub-networks across different cohorts by filtering based on edge weights and PreLect model weights. BacNex then facilitates the visualization of these defined sub-networks.

![alt](img/img3.png?raw=true)
**Fig 3.** The visulized sub-network by BacNex based on [visNetwork](https://github.com/datastorm-open/visNetwork).


#### *Pathway analysis*

To determine which pathways the species in the defined critical sub-network participate in, we employ a sub-network-based analysis method. Users can set the adjusted p-value threshold for Fisher's exact test, which determines the stringency of pathway analysis pipeline.



#### *Network analysis*

BacNex offers network analysis to identify significant sub-networks based on clustering information, enabling you to determine which taxa clusters may be favored by specific cohorts. Additionally, it allows you to visualize the taxonomy clusters of interest and employee the pathway analysis.

## Liscence
This code is made available under the CC BY-NC 4.0 license for non-commercial research applications. For more details, see the [LICENSE](/LICENSE) file.

## Citing Resources
Coming soon ...

## Authors
* Yu Chen Liu
* Yin Cheng Chen